HopsanGUI/HcomHandler.cpp

```diff
@@ -6006,124 +6008,51 @@ void HcomHandler::evaluateExpression(QString expr, VariableType desiredType)
-                return;
-            }
-        }
-
-        if(pData)
-        {
-            mAnsType = HcomHandler::DataVector;
-            mAnsVector = pLogDataHandler->createOrphanVariable(QString("Round%1").arg(pData->getSmartName()), pData->getVariableType());
-            mAnsVector->assignFrom(pData->getSharedTimeOrFrequencyVector(), pData->roundOfData());
-            return;
-        }
-    }
-
-    if(isHcomFunctionCall("floor", expr))
-    {
-        QString argStr = expr.mid(6, expr.size()-7).trimmed();
-        SharedVectorVariableT pData = getLogVariable(argStr);
-        if(!pData)
-        {
-            evaluateExpression(argStr);
-
-            if(mAnsType == HcomHandler::DataVector)
-            {
-                pData = mAnsVector;
-            }
-            else if (mAnsType == HcomHandler::Scalar)
-            {
-                mAnsScalar = qFloor(mAnsScalar);
-                return;
-            }
-        }
-
-        if(pData)
-        {
-            mAnsType = HcomHandler::DataVector;
-            mAnsVector = pLogDataHandler->createOrphanVariable(QString("Floor%1").arg(pData->getSmartName()), pData->getVariableType());
-            mAnsVector->assignFrom(pData->getSharedTimeOrFrequencyVector(), pData->floorOfData());
-            return;
-        }
-    }
-
-    if(isHcomFunctionCall("ceil", expr))
-    {
-        QString argStr = expr.mid(5, expr.size()-6).trimmed();
+    //Simple mathematical vector functions with no arguments
+    FuncMap_t funcMap;
+    funcMap.insert("sin", qMakePair(static_cast<ScalarMathFunction_t>(sin),&VectorVariable::sinOfData));
+    funcMap.insert("cos", qMakePair(static_cast<ScalarMathFunction_t>(cos),&VectorVariable::cosOfData));
+    funcMap.insert("tan", qMakePair(static_cast<ScalarMathFunction_t>(tan),&VectorVariable::tanOfData));
+    funcMap.insert("asin", qMakePair(static_cast<ScalarMathFunction_t>(asin),&VectorVariable::asinOfData));
+    funcMap.insert("acos", qMakePair(static_cast<ScalarMathFunction_t>(acos),&VectorVariable::acosOfData));
+    funcMap.insert("atan", qMakePair(static_cast<ScalarMathFunction_t>(atan),&VectorVariable::atanOfData));
+    funcMap.insert("log", qMakePair(static_cast<ScalarMathFunction_t>(log),&VectorVariable::logOfData));
+    funcMap.insert("exp", qMakePair(static_cast<ScalarMathFunction_t>(exp),&VectorVariable::expOfData));
+    funcMap.insert("sqrt", qMakePair(static_cast<ScalarMathFunction_t>(sqrt),&VectorVariable::sqrtOfData));
+    funcMap.insert("round", qMakePair(static_cast<ScalarMathFunction_t>(round),&VectorVariable::roundOfData));
+    funcMap.insert("floor", qMakePair(static_cast<ScalarMathFunction_t>(floor),&VectorVariable::floorOfData));
+    funcMap.insert("ceil", qMakePair(static_cast<ScalarMathFunction_t>(ceil),&VectorVariable::ceilOfData));
+    funcMap.insert("abs", qMakePair(static_cast<ScalarMathFunction_t>(fabs),&VectorVariable::absOfData));
+    if(desiredType != Scalar && funcMap.contains(getFunctionName(expr))) {
```

Comment:
Previously isHcomFunctionCall("round", expr) and similar was called. That function does additional counting of parenthesis which getFunctionName does not seem to do, unless it solves that problem in some other way.

I dont remember when exactly but at some point I was forced to add counting to avoid treating a sub-part of an expression as a function. It might have been in case of a function taking multiple arguments, where the arguments were them selves expressions taking expressions as arguments. 
Or it could have been to avoid calling a function when arguments where function calls but the wrong number of parenthesis were present

Anyway, I would like that check back somehow, maybe just add it to getFunctionName, so that it returns nothing the parenthesis do not match. I think that should give the same behavior as before.

We really need test for this. I think that I have made a test model / script for these kind of things but I can not find it now. Not sure if those cases were included there though. ("Later" work)

Reply:
I found this comment in isHcomFunctionCall
`// When nOpen == 0 we should have closed the function, and we should be at end of expr, else this is an expression and not a clean function call`
It likely has to do with the interpretation becoming confused when a function call contains sub-expressions with other functions in them. In some way...

Reply:
I solved it the easy way by adding a call to `isHcomFunctionCall` inside `getFunctionName`. It returns an empty string if not a function.